### PR TITLE
Implement http keepalive

### PR DESCRIPTION
**Pull request by gnufied**
This code implements http client connections that are reused between requests. I have verified that it works by running against custom http server that supports keep-alive connections and should work with any http server that supports it.

This pull request should improve performance on a host, where metrics are being emitted too fast (i.e sooner than 10 second interval).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -12,3 +12,4 @@ src/golang.org/*
 .vagrant
 src/fullerite/collector/collector.test
 _vendor/
+fullerite.conf
```
Comment (tsheasha) [inline .gitignore]: This need not be living inside fullerite repository anyways, instead of adding it in .gitignore, you can place it in your home directory.

**Comment by tsheasha** [comment]
May I ask why this is SignalFx specific?

File: src/fullerite/util/http_alive.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	Body       []byte
+	Header     http.Header
+}
+
+// Configure the http connection
+func (connection *HTTPAlive) Configure(timeout time.Duration) {
+	connection.transport = &http.Transport{
+		Dial: (&net.Dialer{
+			Timeout:   timeout,
+			KeepAlive: time.Minute,
```
Comment (tsheasha) [inline src/fullerite/util/http_alive.go]: Setting an explicit KeepAlive time of one minute will not do what we expect it to do. It will keep the connections alive for only one Minute regardless of the emitInterval set for the handler. So if the interval is > 1 minute, it is as if we have no keep-alive. Now I am not sure why someone would want to keep a connection alive for 1 minute but hey, that's not our call to make.

I would recommend setting the value of the KeepAlive to the emitInterval + some constant and to also set MaxIdleConnsPerHost for the Transport and set it to a sane value since we don't want to be creating a lot of idle keep-alive connections without doing some cleanup.

File: src/fullerite/util/http_alive.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+	body, err := ioutil.ReadAll(rsp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	httpAliveResponse := new(HTTPAliveResponse)
+	httpAliveResponse.Body = body
+	httpAliveResponse.StatusCode = rsp.StatusCode
+	httpAliveResponse.Header = rsp.Header
```
Comment (tsheasha) [inline src/fullerite/util/http_alive.go]: Why are we creating a new object for each response that holds some attributes of the response?
Would it not work if we return the response itself?

This will save us some lines of code and hence allow us to merge this with MakeRequest.

File: src/fullerite/util/http_alive.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		return nil, err
+	}
+
+	// Apply user provided headers
+	for key, value := range connection.customHeader {
+		req.Header.Set(key, value)
+	}
+
+	return connection.submitRequest(req)
+}
```
Comment (tsheasha) [inline src/fullerite/util/http_alive.go]: This function `MakeRequest` looks like a wrapper around http.newRequest with no real logic in it, why can't we merge it with `SubmitRequest` and hence make the function live up to it's name of making the request. I am really confused by this split up. And moving around objects between functions unless done with an apparent gain is harmful to maintainers and to performance as well.

File: .gitignore (hunk 1 above)
Comment (gnufied) [inline .gitignore]: ok

File: src/fullerite/util/http_alive.go (hunk 2 above)
Comment (gnufied) [inline src/fullerite/util/http_alive.go]: We can make this configurable I think and yeah I will add support for MaxIdleConnections.

File: src/fullerite/util/http_alive.go (hunk 4 above)
Comment (gnufied) [inline src/fullerite/util/http_alive.go]: I am like a huge fan of concern separation, :D

and `MakeRequest` has two concerns: 

1. Create the http request object, set headers and possibly handle errors if request is invalid.
2. submit that request object to http server, recieve response and handle response errors etc.

Now the golang's `net/http` has chosen to expose above two concerns as two separate function calls - `NewRequest` and `Do` , but I feel that for ease of use in our case we can merge this in one. Internally however, we still have two different things to do, create request object and submit&process response and separate them is a good thing IMO. 

Can you tell me what is confusing about the split up? It seems to be very simply done along the concerns I mentioned above.

File: src/fullerite/util/http_alive.go (hunk 3 above)
Comment (gnufied) [inline src/fullerite/util/http_alive.go]: Since this code now uses `KeepAlive` - the response body must be fully read before next request can be made. Handling response object from `net/http` to the user will force him to handle this problem and she will become responsible for fully reading the body.  

I don't want that to happen and I want to ensure that once `MakeRequest` method returns a response, user can be ready to make request and not worry about what happens with body object.

**Comment by gnufied** [comment]
Also, first I implemented this for sfx handler because - we have to start somewhere and I didn't wanted this to be a huge change. However since most of code is written as reusable interface it can be used from other places where we are making frequent http calls.

File: src/fullerite/handler/signalfx.go [hunk 5]
```diff
@@ -16,10 +14,15 @@ import (
...
 	BaseHandler
-	endpoint  string
-	authToken string
+	endpoint   string
+	authToken  string
+	httpClient *util.HTTPAlive
 }
 
+const (
+	keepAliveGracePeriod = 60
```
Comment (tsheasha) [inline src/fullerite/handler/signalfx.go]: This is a lot of time for a grace period, if i understand correctly this is the time we give to the handler in case it is sending a big payload and it goes a little bit over it's emit interval. Well if it goes 60 seconds over it's emit interval then something is really wrong, we neen't allow that much grace.

File: src/fullerite/util/http_alive.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+import (
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+)
+
+const (
+	maxIdleConnections = 5
```
Comment (tsheasha) [inline src/fullerite/util/http_alive.go]: why 5? why not more or less?

Comment (gnufied) [inline src/fullerite/util/http_alive.go]: So the default is 2. i.e - golang configures maxidelconnection per host to be 2. There is not particular reason honestly why I chose 5. 2 might be too small and anything bigger might be too big.

File: src/fullerite/util/http_alive.go (hunk 3 above)
Comment (tsheasha) [inline src/fullerite/util/http_alive.go]: Isn't it true, since you deferred the function at the top this function will be performed after you return however in the same thread hence will be blocking,. so the user will anyways see the same behaviour.

**Comment by tsheasha** [comment]
other than my last 2 comments nothing more form me, fix and ship +1

File: src/fullerite/util/http_alive.go (hunk 6 above)
Comment (tsheasha) [inline src/fullerite/util/http_alive.go]: I think we need to experiment with this and find a good default value and make this configurable then

File: src/fullerite/handler/signalfx.go (hunk 5 above)
Comment (gnufied) [inline src/fullerite/handler/signalfx.go]: So the default value in golang is 30 seconds - https://github.com/golang/go/blob/master/src/net/http/transport.go#L36

And this graceperiod is unrelated to request timeout we are specifying (which is 2 second btw). This is KeepAlive interval and I think has nothing to do with timeout.

**Comment by poros** [comment]
fix T's comments and ship-it!

**Comment by gnufied** [comment]
@tsheasha @poros I have made both keepaliveinterval and maxidleconnectionsperhost completely configurable via `fullerite.conf`. 

The default values though are same as those provided by golang out of box, which is 30second for keep alive interval and 2 maxidleconnections.

We can run some experiments in production and figure out more accurate values, but I think this should work.

**Comment by tsheasha** [comment]
lgtm +1
